File: solutions/python/saddle-points/3/saddle_points.py
"""
Module for identifying saddle points in a matrix.

A saddle point is an element which is the largest in its row and the smallest in its column.
This module provides a function to identify all such points in a given matrix.

"""


def saddle_points(matrix):
    """
    Identify all saddle points in the given matrix. A saddle point is an element which is
    the largest in its row and the smallest in its column.

    Args:
        matrix (list of list of int): A 2D list representing the matrix to be evaluated.

    Raises:
        ValueError: If the input matrix is irregular (rows are not of the same length).

    Returns:
        list of dict: A list of dictionaries where each dictionary represents a saddle point,
                      containing keys 'row' and 'column' with 1-based indices.
    """
    if any(len(row) != len(matrix[0]) for row in matrix):
        raise ValueError("irregular matrix")

    row_max_positions = (
        (int(row_index), int(col_index))
        for row_index, row in enumerate(matrix)
        for col_index in range(len(row))
        if row[col_index] == max(row)
    )

    return [
        {"row": position[0] + 1, "column": position[1] + 1}
        for position in row_max_positions
        if matrix[position[0]][position[1]]
        == min(
            matrix[row_index][position[1]] for row_index in range(len(matrix))
        )
    ]

File: solutions/javascript/saddle-points/1/saddle-points.js
/**
 * Finds the saddle points in a matrix.
 * A saddle point is an element of the matrix that is the minimum in its column and the maximum in its row.
 *
 * @param {number[][]} matrix - The input matrix.
 * @returns {Object[]} - An array of objects representing the saddle points with their row and column indices.
 */
function saddlePoints(matrix) {
  const maximumRowValues = getMaximumRowValues(matrix);
  const minimumColumnValues = getMinimumColumnValues(matrix);
  return findSaddlePoints(maximumRowValues, minimumColumnValues);
}

/**
 * Gets the maximum values of each row in the matrix.
 *
 * @param {number[][]} matrix - The input matrix.
 * @returns {number[]} - An array of maximum values for each row.
 */
function getMaximumRowValues(matrix) {
  return matrix.map(row => Math.max(...row));
}

/**
 * Gets the minimum values of each column in the matrix.
 *
 * @param {number[][]} matrix - The input matrix.
 * @returns {number[]} - An array of minimum values for each column.
 */
function getMinimumColumnValues(matrix) {
  const minimumColumnValues = [];
  for (let i = 0; i < matrix[0].length; i++) {
    let minimumColumnValue = Number.MAX_VALUE;
    for (let j = 0; j < matrix.length; j++) {
      minimumColumnValue = Math.min(minimumColumnValue, matrix[j][i]);
    }
    minimumColumnValues.push(minimumColumnValue);
  }
  return minimumColumnValues;
}

/**
 * Finds the saddle points given the maximum row values and minimum column values.
 *
 * @param {number[]} maximumRowValues - The maximum values of each row.
 * @param {number[]} minimumColumnValues - The minimum values of each column.
 * @returns {Object[]} - An array of objects representing the saddle points with their row and column indices.
 */
function findSaddlePoints(maximumRowValues, minimumColumnValues) {
  const resultPoints = [];
  for (let i = 0; i < maximumRowValues.length; i++) {
    for (let j = 0; j < minimumColumnValues.length; j++) {
      if (maximumRowValues[i] === minimumColumnValues[j]) {
        resultPoints.push({ row: i + 1, column: j + 1 });
      }
    }
  }
  return resultPoints;
}

export { saddlePoints };